### PR TITLE
[Merged by Bors] - Used p2p peers in p2p gossip protocol, big refactor.

### LINKED ISSUE #2017: Use `p2p.Peers` in `gossip.Protocol`
## Description

We implemented the peer event handling code in `gossip.Protocol` before we wrote `p2p.Peers`, so there's much duplication there. we want to reduce that by making `gossip.Protocol` use the `p2p.Peers` struct to get informed about connected peers.

## Affected code
- `p2p.Peers`
- `p2p.gossip.Protocol`

## Notes:
this might require refactoring `p2p.Peers` to a different package (under `p2p`) to prevent a dependency cycle.

### PR DESCRIPTION
**Pull request by anton-kuklin**
## Motivation
Closes #2017 

## Changes
Used p2p/peers in p2p/gossip protocol, instead of custom implementation. Rewrote gossip protocol unit tests. Added some cosmetic changes.

## Test Plan
Ran p2p package unit tests, peers test and rewrote gossip tests

### CODE REVIEW

**Comment by anton-kuklin** [comment]
Will fix tests

**Comment by anton-kuklin** [comment]
Was some sort of CI bug :) @y0sher  ready for review

File: p2p/gossip/protocol.go [hunk 1]
```diff
@@ -2,20 +2,29 @@
...
 	"github.com/spacemeshos/go-spacemesh/p2p/config"
 	"github.com/spacemeshos/go-spacemesh/p2p/metrics"
 	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+	"github.com/spacemeshos/go-spacemesh/p2p/peers"
 	"github.com/spacemeshos/go-spacemesh/p2p/service"
 	"github.com/spacemeshos/go-spacemesh/priorityq"
-	"sync"
 )
 
+//go:generate mockgen -package=gossip -destination=./protocol_mock_test.go -source=./protocol.go peersManager, baseNetwork, prioQ
```
Comment (y0sher) [inline p2p/gossip/protocol.go]: can this be moved to `protocol_test.go`?

Comment (anton-kuklin) [reply p2p/gossip/protocol.go]: sure, guess it will be better

**Review by y0sher — APPROVED** [review]
good job rewriting the tests and refactoring. thanks!

**Comment by y0sher** [comment]
bors merge

**Comment by anton-kuklin** [comment]
@y0sher may we restart?)

**Comment by y0sher** [comment]
@justplesh , yes sorry we had some ci issues. I hope it will be better now.
bors merge

**Comment by anton-kuklin** [comment]
Hm, out of range looks very strange, failing on shuffle.
https://github.com/spacemeshos/go-spacemesh/blob/5282bba4ccce31865e5b4902bde6a6af7e48ecdb/p2p/peers/peers.go#L52

Will investigate.

**Comment by y0sher** [comment]
let me know if there's trouble and you need help.

**Comment by anton-kuklin** [comment]
@y0sher so, I did some investigations
I found that spacemesh rand package uses rand.NewSource as rand.New argument 
https://github.com/spacemeshos/go-spacemesh/blob/b9fdeb9b230a607fa36e6e83225971de07cf1087/rand/rand.go#L12

As the official documentation says: 

> Unlike the default Source used by top-level functions, this source is not safe for concurrent use by multiple goroutines.

https://golang.org/pkg/math/rand/#NewSource

As a result, all concurrent usages of spacemesh rand package have race conditions, which luckily happened during the last unit tests.

If I am wrong - please correct me.

**Comment by y0sher** [comment]
I can see the problem you are talking about with concurrently using rand. though I don't see how this would've caused the error we are seeing.

**Comment by y0sher** [comment]
bors try

**Comment by anton-kuklin** [comment]
@y0sher 

> though I don't see how this would've caused the error we are seeing.

For example, rand.Shuffle is used concurrently, and the case is accessing indexes from other concurrently randomized value, that is out of bound of an array. This is a very huge bug and we need to fix it everywhere with rand. Rand package can't be used in such a way when concurrent access is possible. I guess the logic must be changed. We can implement it in another issue, but it is very important, we can't just ignore it. It's not only about p2p using rand, but it's also a problem in general.

**Comment by y0sher** [comment]
I see. you are right this is a broader issue. 
from golangs github -
```
If you need to get randomness from multiple goroutine,
the solution is to instantiate a random generator for
each goroutine.
```

so to solve current situation you can create a new rand and source for each `Peers` instance. this is the recommended usage pattern and it'll prevent these issue at least in `Peers` so we can conclude this PR.

**Comment by anton-kuklin** [comment]
Exactly, but I guess we need to create an issue and validate all usages of rand for sequential rand usage.

**Comment by anton-kuklin** [comment]
@y0sher used separate rand with Mutex - check please and let's run bors

**Comment by y0sher** [comment]
I don't think the mutex is needed since `Peers` is created for each component and there's no use as of now that we do of `GetPeers` concurrently. let's drop the mutex and I'll run bors

**Comment by anton-kuklin** [comment]
hm, yep you are right, but let's leave a comment, that method is concurrently not safe

**Comment by y0sher** [comment]
bors merge

**Pull request closed without merging**